sbin/dumpon/dumpon.8
.\" Copyright (c) 1980, 1991, 1993
.\"	The Regents of the University of California.  All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. Neither the name of the University nor the names of its contributors
.\"    may be used to endorse or promote products derived from this software
.\"    without specific prior written permission.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
.\" ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
.\" FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
.\" DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
.\" OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
.\" HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
.\" LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
.\" OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
.\" SUCH DAMAGE.
.\"
.\"     From: @(#)swapon.8	8.1 (Berkeley) 6/5/93
.\" $FreeBSD: src/sbin/dumpon/dumpon.8,v 1.11.2.12 2003/01/26 03:12:04 keramida Exp $
.\"
.Dd September 23, 2018
.Dt DUMPON 8
.Os
.Sh NAME
.Nm dumpon , dumpoff
.Nd "specify a device for crash dumps"
.Sh SYNOPSIS
.Nm
.Ar special_file
.Nm
.Cm off
.Nm dumpoff
.Sh DESCRIPTION
The
.Nm
utility is used to specify a device where the kernel can save a crash dump in
the case of a panic.
.Pp
Calls to
.Nm
normally occur from the system multi-user initialization file
.Pa /etc/rc ,
controlled by the
.Dq dumpdev
variable in the boot time configuration file
.Pa /etc/rc.conf .
.Pp
The default type of kernel crash dump is the mini crash dump.
Mini crash dumps hold only memory pages in use by the kernel.
Alternatively, full memory dumps can be enabled by setting the
.Va debug.minidump
.Xr sysctl 8
variable to 0.
.Pp
For most systems the size of the specified dump device must be at least
the size of physical memory.
Even though an additional header is added to the dump,
the BIOS for a platform typically holds back some memory, so it is not usually
necessary to size the dump device larger than the actual amount of RAM
available in the machine.
.Sh NOTES
Since a
.Xr panic 9
condition may occur in a situation
where the kernel cannot trust its internal representation
of the state of any given file system,
one of the system swap devices,
and
.Em not
a device containing a file system,
should be used as the dump device.
.Pp
The
.Nm
utility operates by setting the
.Xr sysctl 3
MIB variable
.Va kern.dumpdev
to the device number of the designated
.Ar special_file
or to
.Dv NODEV
(meaning that no dumps are to be taken) if the utility is invoked as
.Nm dumpoff
or if
.Ar special_file
is the text string:
.Dq Li off .
.Pp
In order to change the device for crash dumps, one needs to run the
.Nm dumpoff
utility to disable the current dump device before configuring the new one.
.Pp
Since
.Nm
cannot be used during kernel initialization, the
.Va dumpdev
variable of
.Xr loader 8
must be used to enable dumps for system panics which occur
during kernel initialization.
.Sh FILES
.Bl -tag -width "/dev/{ad,da}?s?b" -compact
.It Pa /dev/{ad,da}?s?b
standard swap areas
.It Pa /etc/rc.conf
boot-time system configuration
.El
.Sh SEE ALSO
.Xr sysctl 3 ,
.Xr fstab 5 ,
.Xr rc.conf 5 ,
.Xr init 8 ,
.Xr loader 8 ,
.Xr rc 8 ,
.Xr savecore 8 ,
.Xr swapon 8 ,
.Xr panic 9
.Sh HISTORY
The
.Nm
utility appeared in
.Fx 2.0.5 .
The
.Nm dumpoff
utility first appeared in
.Dx 5.3 .
.Sh BUGS
Because the file system layer is already dead by the time a crash dump
is taken, it is not possible to send crash dumps directly to a file.


sbin/dumpon/dumpon.c
/*
 * Copyright (c) 1980, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. Neither the name of the University nor the names of its contributors
 *    may be used to endorse or promote products derived from this software
 *    without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
 * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
 * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
 * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
 * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
 * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
 * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
 * SUCH DAMAGE.
 *
 * @(#)swapon.c	8.1 (Berkeley) 6/5/93
 * $FreeBSD: src/sbin/dumpon/dumpon.c,v 1.10.2.2 2001/07/30 10:30:05 dd Exp $
 */

#include <err.h>
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <string.h>
#include <unistd.h>
#include <fstab.h>
#include <sys/param.h>
#include <sys/sysctl.h>
#include <sys/stat.h>
#include <sysexits.h>

static const char *sc_name = "kern.dumpdev";

static struct stat *get_dumpdev(void);
static void usage(void) __dead2;

int
main(int argc, char **argv)
{
	int ch;
	struct stat stab;
	struct stat *stab_old;
	char *path, *p;
	bool is_dumpoff;

	if (strstr((p = strrchr(argv[0], '/')) ? p+1 : argv[0],
		   "dumpoff") != 0) {
		is_dumpoff = true;
	} else {
		is_dumpoff = false;
	}

	while ((ch = getopt(argc, argv, "v")) != -1) {
		switch (ch) {
		case 'v':
			/* backward compatibility only */
			break;
		case '?':
		default:
			usage();
		}
	}
	argc -= optind;
	argv += optind;

	if ((is_dumpoff && argc != 0) || (!is_dumpoff && argc != 1))
		usage();

	path = argv[0];
	if (is_dumpoff || strcmp(path, "off") == 0) {
		stab.st_rdev = NODEV;
	} else {
		path = getdevpath(path, 0);
		if (stat(path, &stab) != 0)
			err(EX_OSFILE, "%s", path);

		if (!S_ISCHR(stab.st_mode)) {
			errx(EX_USAGE,
			     "%s: must specify a character disk device",
			     path);
		}
	}

	stab_old = get_dumpdev();

	if (stab.st_rdev == stab_old->st_rdev) {
		if (stab.st_rdev == NODEV) {
			printf("dumpon: crash dumps already disabled.\n");
		} else {
			printf("dumpon: crash dumps already configured "
			       "to the given device.\n");
		}
	} else if (stab.st_rdev == NODEV || stab_old->st_rdev == NODEV) {
		if (sysctlbyname(sc_name, NULL, NULL,
				 &stab.st_rdev, sizeof stab.st_rdev) != 0) {
			err(EX_OSERR, "sysctl: %s", sc_name);
		}
		if (stab.st_rdev == NODEV) {
			printf("dumpon: crash dumps disabled\n");
		} else {
			printf("dumpon: crash dumps to %s (%lu, %#lx)\n",
			       path,
			       (unsigned long)major(stab.st_rdev),
			       (unsigned long)minor(stab.st_rdev));
		}
	} else {
		warnx("crash dumps already configured "
		      "to another device (%lu, %#lx)",
		      (unsigned long)major(stab.st_rdev),
		      (unsigned long)minor(stab.st_rdev));
		errx(EX_USAGE, "you need to run 'dumpoff' first.");
	}

	return 0;
}


static struct stat *
get_dumpdev(void)
{
	struct stat *stab;
	size_t len;

	if ((stab = malloc(sizeof(*stab))) == NULL)
		err(EX_OSERR, "malloc");

	memset(stab, 0, sizeof(*stab));
	len = sizeof(stab->st_rdev);
	if (sysctlbyname(sc_name, &stab->st_rdev, &len, NULL, 0) != 0)
		err(EX_OSERR, "sysctl: %s", sc_name);

	return stab;
}

static void
usage(void)
{
	fprintf(stderr,
		"usage: dumpon [-v] special_file\n"
		"       dumpon [-v] off\n"
		"       dumpoff [-v]\n");
	exit(EX_USAGE);
}
